Readme.markdown
ObsidianDestroyer plugin v2.3.1<br>
by Pandemoneus<br>
https://github.com/Pandemoneus<br>

Official Maintained<br>
https://github.com/drtshock/ObsidianDestroyer

How to install:
----------------
1. Copy 'ObsidianDestroyer.jar' into your 'plugins/' folder.
2. Load your server
3. Edit the newly created 'config.yml' in 'plugins/ObsidianDestroyer/' and set your preferences

How to uninstall:
-----------------
1. Delete the folder called 'ObsidianDestroyer' in plugins/.
2. Delete 'ObsidianDestroyer.jar'

Options:
-----------------
Radius - determines how far around the origin of the explosion Obsidian can be destroyed (default: 3) WARNING: High values probably cause lag<br>
FluidsProtect - set to false if you want obsidan to be able to hit by explosions from within water (default: true)<br>
EnabledFor.TNT - set true if TNT is allowed to destroy Obsidian (default: true)<br>
EnabledFor.Cannons - set true if you would like the cannons plugin to break obsidian (default: false) https://github.com/DerPavlov/Cannons/<br>
EnabledFor.Creepers - set true if Creepers are allowed to destroy Obsidian (default: false)<br>
EnabledFor.Ghasts - set true if Ghasts are allowed to destroy Obsidian (default: false)<br>
Durability.Enabled - set true if you want to use the durability feature (default: false)<br>
Durability.Amount - determines after how many TNT explosions in the radius the Obsidian block gets destroyed (default: 1)<br>
Durability.ResetEnabled - set true if you want to reset the durability back to max after a certain time has passed (default: true)<br>
Durability.ResetTime - time in milliseconds that has to pass before the durability is reset (default: '600000' (10 minutes))<br>
Blocks.ChanceToDrop - set the chance to drop an obsidian block when it was blown up, set to 1.0 to always drop a block (default: 0.7 (70%))


Permission nodes:
-----------------
obsidiandestroyer.help //makes help command available<br>
obsidiandestroyer.config.reload //makes reload command available<br>
obsidiandestroyer.config.info //makes info command available<br>
obsidiandestroyer.durability.reset //makes reset command available

Commands:
-----------------
obsidiandestroyer (alias: od) - shows the help<br>
obsidiandestroyer reload (alias: od reload) - reloads the plugin<br>
obsidiandestroyer info (alias: od info) - shows the currently loaded config<br>
obsidiandestroyer reset (alias: od reset) - resets all currently saved obsidian durabilities


Changelog.txt
Version 2.3.1
- Added support for tekkit.  Obsidian Microblocks now handled.

Version 2.3
- Added FluidsProtect option into config.  Can enable damage to obsidian in water.
- Corrected bug with plugin version reporting null

Version 2.2
- Corrected Metrics not starting.

Version 2.1.x
- Fluids stop all damage to obsidian

Version 2.1
- Removed Vault dependency.

Version 2.0
- Added 1.4.5 support
- New author / maintainers: drtshock, squidicc

Version 1.3.1
- Corrected issue with Cannons plugin support drop

Version 1.3
- Added metrics support

Version 1.25
- Added Cannons plugin support

Version 1.21
- Added echantment Tables

Version 1.1
- Updated for minecraft 1.2

Version 1.052
- Added chance to drop Obsidian blocks

Version 1.051
- Fixed NullPointerException in config

Version 1.05
- Durability can now restore back to max after a certain time passed

Version 1.04
- Durability now saves when the server reloads/restarts

Version 1.03
- Added durability

Version 1.02
- Added Permissions
- Added aliases
- Increased maintainability

Version 1.01
- Added command to reload config ingame
- Added command to show loaded config ingame
- Added check to not mess with plugins that prevent explosions

Version 1.0
- Initial release


src/com/pandemoneus/obsidianDestroyerTekkit/ODConfig.java
package com.pandemoneus.obsidianDestroyerTekkit;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.HashMap;

import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.Plugin;

/**
 * The configuration file for the ObsidianDestroyer plugin, uses YML.
 * 
 * @author Pandemoneus
 * 
 */
public final class ODConfig {

	private ObsidianDestroyer plugin;
	private static String pluginName;
	private static String pluginVersion;

	/**
	 * File handling
	 */
	private static String directory = "plugins" + File.separator + ObsidianDestroyer.getPluginName() + File.separator;
	private File configFile = new File(directory + "config.yml");
	private File durabilityFile = new File(directory + "durability.dat");
	private YamlConfiguration bukkitConfig = new YamlConfiguration();

	/**
	 * Default settings
	 */
	private int explosionRadius = 3;
	private boolean tntEnabled = true;
	private boolean cannonsEnabled = false;
	private boolean creepersEnabled = false;
	private boolean ghastsEnabled = false;
	private boolean durabilityEnabled = false;
	private int durability = 1;
	private boolean durabilityTimerEnabled = true;
	private long durabilityTime = 600000L; // 10 minutes
	private double chanceToDropBlock = 0.7;
	private boolean waterProtection = true;

	/**
	 * Associates this object with a plugin
	 * 
	 * @param plugin
	 *            the plugin
	 */
	public ODConfig(ObsidianDestroyer plugin) {
		this.plugin = plugin;
		pluginName = ObsidianDestroyer.getPluginName();
	}

	/**
	 * Loads the configuration used by this plugin.
	 * 
	 * @return true if config loaded without errors
	 */
	public boolean loadConfig() {
		boolean isErrorFree = true;
		pluginVersion = ObsidianDestroyer.getVersion();

		new File(directory).mkdir();

		if (configFile.exists()) {
			try {
				bukkitConfig.load(configFile);

				if (bukkitConfig.getString("Version", "").equals(pluginVersion)) {
					// config file exists and is up to date
					Log.info(pluginName + " config file found, loading config...");
					loadData();
				} else {
					// config file exists but is outdated
					Log.info(pluginName + " config file outdated, adding old data and creating new values. " + "Make sure you change those!");
					loadData();
					writeDefault();
				}
			} catch (Exception e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		} else {
			// config file does not exist
			try {
				Log.info(pluginName + " config file not found, creating new config file...");
				configFile.createNewFile();
				writeDefault();
			} catch (IOException ioe) {
				Log.severe("Could not create the config file for " + pluginName + "!");
				ioe.printStackTrace();
				isErrorFree = false;
			}
		}

		return isErrorFree;
	}

	private void loadData() {
		try {
			bukkitConfig.load(configFile);

			explosionRadius = bukkitConfig.getInt("Radius", 3);
			waterProtection = bukkitConfig.getBoolean("FluidsProtect", true);

			tntEnabled = bukkitConfig.getBoolean("EnabledFor.TNT", true);
			cannonsEnabled = bukkitConfig.getBoolean("EnabledFor.Cannons", false);
			creepersEnabled = bukkitConfig.getBoolean("EnabledFor.Creepers", false);
			ghastsEnabled = bukkitConfig.getBoolean("EnabledFor.Ghasts", false);

			durabilityEnabled = bukkitConfig.getBoolean("Durability.Enabled", false);
			durability = bukkitConfig.getInt("Durability.Amount", 1);
			durabilityTimerEnabled = bukkitConfig.getBoolean("Durability.ResetEnabled", true);
			durabilityTime = readLong("Durability.ResetAfter", "600000");

			chanceToDropBlock = bukkitConfig.getDouble("Blocks.ChanceToDrop", 0.7);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	private void writeDefault() {
		write("Version", ObsidianDestroyer.getVersion());
		write("Radius", explosionRadius);
		write("FluidsProtect", waterProtection);

		write("EnabledFor.TNT", tntEnabled);
		write("EnabledFor.Cannons", cannonsEnabled);
		write("EnabledFor.Creepers", creepersEnabled);
		write("EnabledFor.Ghasts", ghastsEnabled);

		write("Durability.Enabled", durabilityEnabled);
		write("Durability.Amount", durability);
		write("Durability.ResetEnabled", durabilityTimerEnabled);
		write("Durability.ResetAfter", "" + durabilityTime);

		write("Blocks.ChanceToDrop", chanceToDropBlock);

		loadData();
	}

	private void write(String key, Object o) {
		try {
			bukkitConfig.load(configFile);
			bukkitConfig.set(key, o);
			bukkitConfig.save(configFile);
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	/**
	 * Reads a string representing a long from the config file.
	 * 
	 * Returns '0' when an exception occurs.
	 * 
	 * @param key
	 *            the key
	 * @param def
	 *            default value
	 * @return the long specified in 'key' from the config file, '0' on errors
	 */
	private long readLong(String key, String def) {
		try {
			bukkitConfig.load(configFile);
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		// Bukkit Config has no getLong(..)-method, so we are using Strings
		String value = bukkitConfig.getString(key, def);

		long tmp = 0;

		try {
			tmp = Long.parseLong(value);
		} catch (NumberFormatException nfe) {
			Log.warning("Error parsing a long from the config file. Key=" + key);
			nfe.printStackTrace();
		}

		return tmp;
	}

	/**
	 * Returns the explosion radius.
	 * 
	 * @return the explosion radius
	 */
	public int getRadius() {
		return explosionRadius;
	}

	/**
	 * Returns whether TNT is allowed to destroy Obsidian.
	 * 
	 * @return whether TNT is allowed to destroy Obsidian
	 */
	public boolean getTntEnabled() {
		return tntEnabled;
	}
	
	/**
	 * Returns whether Cannons are allowed to destroy Obsidian.
	 * 
	 * @return whether Cannons are allowed to destroy Obsidian
	 */
	public boolean getCannonsEnabled() {
		return cannonsEnabled;
	}


	/**
	 * Returns whether Creepers are allowed to destroy Obsidian.
	 * 
	 * @return whether Creepers are allowed to destroy Obsidian
	 */
	public boolean getCreepersEnabled() {
		return creepersEnabled;
	}

	/**
	 * Returns whether Ghasts are allowed to destroy Obsidian.
	 * 
	 * @return whether Ghasts are allowed to destroy Obsidian
	 */
	public boolean getGhastsEnabled() {
		return ghastsEnabled;
	}

	/**
	 * Returns whether durability for Obsidian is enabled.
	 * 
	 * @return whether durability for Obsidian is enabled
	 */
	public boolean getDurabilityEnabled() {
		return durabilityEnabled;
	}

	/**
	 * Returns the max durability.
	 * 
	 * @return the max durability
	 */
	public int getDurability() {
		return durability;
	}

	/**
	 * Returns whether durability timer for Obsidian is enabled.
	 * 
	 * @return whether durability timer for Obsidian is enabled
	 */
	public boolean getDurabilityResetTimerEnabled() {
		return durabilityTimerEnabled;
	}

	/**
	 * Returns the time in milliseconds after which the durability gets reset.
	 * 
	 * @return the time in milliseconds after which the durability gets reset
	 */
	public long getDurabilityResetTime() {
		return durabilityTime;
	}

	/**
	 * Returns the chance to drop an item from a blown up block.
	 * 
	 * @return the chance to drop an item from a blown up block
	 */
	public double getChanceToDropBlock() {
		return chanceToDropBlock;
	}
	
	/**
	 * Returns whether water stops damage to obsidian
	 * 
	 * @return whether water stops damage to obsidian
	 */
	public boolean getWaterProtection() {
		return waterProtection;
	}

	/**
	 * Returns a list containing all loaded keys.
	 * 
	 * @return a list containing all loaded keys
	 */
	public String[] printLoadedConfig() {
		return new String[] { "this doesn't work." };
	}

	/**
	 * Returns the config file.
	 * 
	 * @return the config file
	 */
	public File getConfigFile() {
		return configFile;
	}

	/**
	 * Returns the associated plugin.
	 * 
	 * @return the associated plugin
	 */
	public Plugin getPlugin() {
		return plugin;
	}

	/**
	 * Saves the durability hash map to a file.
	 */
	public void saveDurabilityToFile() {
		if (plugin.getListener() == null || plugin.getListener().getObsidianDurability() == null) {
			return;
		}

		HashMap<Integer, Integer> map = plugin.getListener().getObsidianDurability();

		new File(directory).mkdir();

		try {
			ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(durabilityFile));
			oos.writeObject(map);
			oos.flush();
			oos.close();
		} catch (IOException e) {
			Log.severe("Failed writing obsidian durability for " + ObsidianDestroyer.getPluginName());
			e.printStackTrace();
		}
	}

	/**
	 * Loads the durability hash map from a file.
	 * 
	 * @return the durability hash map from a file
	 */
	@SuppressWarnings("unchecked")
	public HashMap<Integer, Integer> loadDurabilityFromFile() {
		if (!durabilityFile.exists() || plugin.getListener() == null || plugin.getListener().getObsidianDurability() == null) {
			return null;
		}

		new File(directory).mkdir();

		HashMap<Integer, Integer> map = null;
		Object result = null;

		try {
			ObjectInputStream ois = new ObjectInputStream(new FileInputStream(durabilityFile));
			result = ois.readObject();
			map = (HashMap<Integer, Integer>) result;
			ois.close();
		} catch (IOException ioe) {
			Log.severe("Failed reading obsidian durability for " + ObsidianDestroyer.getPluginName());
			ioe.printStackTrace();
		} catch (ClassNotFoundException cnfe) {
			Log.severe("Obsidian durability file contains an unknown class, was it modified?");
			cnfe.printStackTrace();
		}

		return map;
	}
}
